lib/livebook/zta/tailscale.ex

```diff
@@ -0,0 +1,103 @@
+defmodule Livebook.ZTA.Tailscale do
+  @doc """
+  To integrate Tailscale authentication with Livebook,
+  set the `LIVEBOOK_IDENTITY_PROVIDER` environment variable to `tailscale:tailscale-socket-path`.
+
+  If you want to access the Livebook on the same machine as you are hosting it on,
+  you will also need to set the LIVEBOOK_IP variable to your Tailscale IP.
+
+  To do both of these things, run
+
+  ```bash
+  LIVEBOOK_IP=$(tailscale ip | head -1 | tr -d '\n') LIVEBOOK_IDENTITY_PROVIDER=tailscale:/var/run/tailscale/tailscaled.sock livebook server
+  ```
+
+  See https://tailscale.com/blog/tailscale-auth-nginx/ for more information
+  on how Tailscale authorization works.
+
+  ## MacOS
+
+  On MacOS, when Tailscale is installed via the Mac App Store, no unix socket is exposed.
+  Instead, a TCP port is made available, protected via a password, which needs to be located.
+  Tailscale itself uses lsof for this. This method is replicated in the bash script below,
+  which will start Livebook with your Tailscale IP and correct port and password.
+
+  ```bash
+  #!/bin/bash
+  addr_info=$(lsof -n -a -c IPNExtension -F | sed -n 's/.*sameuserproof-\([[:digit:]]*-.*\).*/\1/p')
+  port=$(echo "$addr_info" | cut -d '-' -f 1)
+  pass=$(echo "$addr_info" | cut -d '-' -f 2)
+  LIVEBOOK_IP=$(exec $(ps -xo comm | grep MacOS/Tailscale$) ip | head -1 | tr -d '\n') LIVEBOOK_IDENTITY_PROVIDER=tailscale:http://:$pass@127.0.0.1:$port livebook server
+  ```
+  """
+
+  use GenServer
+  require Logger
+
+  defstruct [:name, :address]
+
+  def start_link(opts) do
+    options = [address: opts[:identity][:key]]
+    GenServer.start_link(__MODULE__, options, name: opts[:name])
+  end
+
+  def authenticate(name, conn, _) do
+    remote_ip = to_string(:inet_parse.ntoa(conn.remote_ip))
+    user = GenServer.call(name, {:authenticate, remote_ip})
+    {conn, user}
+  end
+
+  @impl true
+  def init(options) do
+    state = struct!(__MODULE__, options)
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_call({:authenticate, remote_ip}, _from, state) do
+    user = authenticated_ip(remote_ip, state)
+    {:reply, user, state}
+  end
+
+  defp authenticated_ip(remote_ip, %{address: address} = _state) do
+    [url, options] =
+      if String.starts_with?(address, "/") do
+        # Assume address starting with a slash is a Unix socket
+        [
+          "http://local-tailscaled.sock/localapi/v0/whois?addr=#{remote_ip}:1",
+          [
+            unix_socket: address,
+            # Req or Finch do not pass on the host from the URL when using a unix socket,
+            # so we set the host header explicitly
+            headers: [host: "local-tailscaled.sock"]
+          ]
+        ]
+      else
+        uri = URI.parse(address)
+
+        options =
+          if uri.userinfo do
+            # Req does not handle userinfo
+            [headers: [authorization: "Basic #{Base.encode64(uri.userinfo)}"]]
```

Comment:
You may be able to use the `basic: uri.userinfo` option. :)

Reply:
Is that a Req option? I can't find it in the Req codebase

Reply:
Sorry, it might be `auth: {:basic, uri.user_info}`?

Reply:
neither is supported. We can do auth: {user, pass}. In hindsight {:basic, userinfo} is not a bad idea.

Reply:
I changed it to `[auth: "Basic #{Base.encode64(uri.userinfo)}"]` which saves a few characters, but I think that's the best we can do right now :)